src/main/java/gregtech/api/pipelike/PipeLikeObjectFactory.java

```diff
@@ -0,0 +1,298 @@
+        if (material instanceof GemMaterial) {
+            return net.minecraft.block.material.Material.ROCK;
+        } else if (material instanceof IngotMaterial) {
+            if (material.toString().contains("rubber") || material.toString().contains("poly") || material.toString().contains("plastic"))
+                return net.minecraft.block.material.Material.SPONGE;
+            return net.minecraft.block.material.Material.IRON;
+        } else if (material.toString().contains("wood")) {
+            return net.minecraft.block.material.Material.WOOD;
+        }
+        return net.minecraft.block.material.Material.ROCK;
+    }
+
+    protected BlockPipeLike<Q, P, T, C> createBlock(Material material, P[] materialProperties) {
+        return new BlockPipeLike<>(this, getBlockMaterial(material), material, materialProperties);
+    }
+
+    protected abstract void initBlock(BlockPipeLike<Q, P, T, C> block);
+    protected abstract P createActualProperty(Q baseProperty, P materialProperty);
+
+    @SuppressWarnings("unchecked")
+    public BlockPipeLike<Q, P, T, C> createBlock(Material material, P materialProperty) {
+        P[] materialProperties = (P[]) Array.newInstance(classTileProperty, baseProperties.length);
+        for (int i = 0; i < baseProperties.length; i++) {
+            materialProperties[i] = createActualProperty(baseProperties[i], materialProperty);
+        }
+        BlockPipeLike<Q, P, T, C> block = createBlock(material, materialProperties);
+        block.setUnlocalizedName(name);
+        initBlock(block);
+        blocks.put(material, block);
+        return block;
+    }
+
+    public TileEntityPipeLike<Q, P, T, C> createNewTileEntity() {
+        return new TileEntityPipeLike<>(this);
+    }
+
+    @SuppressWarnings("unchecked")
+    public T getTile(IBlockAccess world, BlockPos pos) {
+        TileEntity tile = world.getTileEntity(pos);
+        if (tile == null) return null;
+        if (classTileInterface.isInstance(tile)) {
+            return (T) tile;
+        } else if (Loader.isModLoaded(GTValues.MODID_FMP)) {
+            return getFromMultipart(tile);
+        }
+        return null;
+    }
+
+    public BlockPipeLike<Q, P, T, C> getBlock(Material material) {
+        return blocks.get(material);
+    }
+
+    public Q getBaseProperty(int index) {
+        return index >= 0 && index < baseProperties.length ? baseProperties[index] : null;
+    }
+
+    /////////////////////////////// MULTIPART METHODS //////////////////////////////////////
+
+    @Optional.Method(modid = GTValues.MODID_FMP)
+    private static void registerMultipartFactory() {
+        MultiPartRegistry.registerParts((IPartFactory) (identifier, client) -> {
+            PipeLikeObjectFactory factory = multipartFactories.get(identifier);
+            if (factory == null) return null;
+            return factory.createMultipart();
+        }, multipartFactories.keySet());
+        MultiPartRegistry.registerConverter(new IPartConverter() {
```

Comment:
Please don't use such big nested anonymous classes. It affects visibility quite bad. Better move these into some dedicated place.

Reply:
Got it.